Add exposures and outcomes properties to graph models

MAVRICK-1 opened the pull request:
Changes:
- Add exposures and outcomes properties to _GraphRolesMixin with getter/setter methods
- Extend AncestralBase.__init__ to accept exposures and outcomes parameters
- Update MAG.__init__ to support the new parameters
- Add comprehensive tests for the new functionality
- Maintain backward compatibility with existing code

The implementation allows users to easily assign and manage exposure and outcome roles in causal graphs using an intuitive property-based approach, making it easier to define frequently used node roles in causal inference workflows.

Fixes #2379

MAVRICK-1 committed:
Add exposures and outcomes properties to graph models

This commit implements the feature requested in issue #2379 by adding
exposures and outcomes properties to causal graph models, following
the same pattern as the existing latents property.

Changes:
- Add exposures and outcomes properties to _GraphRolesMixin with getter/setter methods
- Extend AncestralBase.__init__ to accept exposures and outcomes parameters
- Update MAG.__init__ to support the new parameters
- Add comprehensive tests for the new functionality
- Maintain backward compatibility with existing code

The implementation allows users to easily assign and manage exposure
and outcome roles in causal graphs using an intuitive property-based
approach, making it easier to define frequently used node roles in
causal inference workflows.

Fixes #2379

MAVRICK-1 committed:
Apply black formatting to maintain code style

- Reformatted MAG.py super().__init__ call for better readability
- Reformatted test assertions for better line length compliance
- All pre-commit hooks now pass successfully

daehyun99 commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -222,3 +222,87 @@ def observed(self):
+        Create a DAG with exposures and check the exposures value.
+
+        >>> from pgmpy.base import DAG
+        >>> G = DAG(ebunch=[("a", "b")], roles={"exposure": "a"})
```
I think that modifying the example code for granting the `exposure` role to use the `exposures` property will make it easier for users to use the feature. (Please refer to the example for the `latents` property)

https://github.com/pgmpy/pgmpy/blob/08ed278d84ca1d235ab3485ff0db39eb795bcaf6/pgmpy/base/_mixin_roles.py#L174-L177

daehyun99 commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -222,3 +222,87 @@ def observed(self):
+        Create a DAG with outcomes and check the outcomes value.
+
+        >>> from pgmpy.base import DAG
+        >>> G = DAG(ebunch=[("a", "b")], roles={"outcome": "b"})
```
I think that the example code for `outcome` would be good to modify, similar to the example code for `latents`.

daehyun99 commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -222,3 +222,87 @@ def observed(self):
+            self.without_role(
+                role="exposure", variables=self.get_role("exposure"), inplace=True
+            )
+        self.with_role(role="exposure", variables=variables, inplace=True)
```
I think it would be better to change the role name from `exposure` to `exposures`, although it's a small difference. This is because I think consistency with `latents` will be more intuitive for users.

daehyun99 commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -222,3 +222,87 @@ def observed(self):
+            self.without_role(
+                role="outcome", variables=self.get_role("outcome"), inplace=True
+            )
+        self.with_role(role="outcome", variables=variables, inplace=True)
```
I think it would be better if `outcome` were also changed to `outcomes`.

MAVRICK-1 commented:
@daehyun99 @ankurankan  can I get a review on this

daehyun99 reviewed (commented):
Hi @MAVRICK-1 
Thank you for the PR.

I'm not the code reviewer, but I have experience doing similar work before, so I will leave comments on parts that might be good to modify.

The other things look good to me.

ankurankan reviewed (changes requested):
@MAVRICK-1 Thanks for the PR. I think the PR looks good once @daehyun99's comments are addressed.

MAVRICK-1 committed:
Merge branch 'dev' into feature/add-exposures-outcomes-properties

MAVRICK-1 committed:
Address code review feedback: update examples and role naming consistency

- Update exposures property example to use exposures parameter instead of roles
- Update outcomes property example to use outcomes parameter instead of roles
- Change internal role names from 'exposure'/'outcome' to 'exposures'/'outcomes' for consistency with latents
- Update corresponding test assertions to match new role names
- All role-related tests pass (13/13)

MAVRICK-1 committed:
Fix test failures: update all test files to use new role names

- Update test_causal_graph.py to use 'exposures'/'outcomes' instead of 'exposure'/'outcome'
- Update test_adjustment.py to use new role names in all DAG fixtures and assertions
- Update test_base.py to use new role names in identification tests
- Update test_frontdoor.py to use new role names in frontdoor identification tests
- Apply black formatting to updated files

MAVRICK-1 committed:
Fix identification modules: update to use new role names

- Update adjustment.py to use 'exposures'/'outcomes' instead of 'exposure'/'outcome'
- Update frontdoor.py to use new role names in identification algorithms
- Update base.py to use new role names in examples
- Fixes NotImplementedError and IndexError in identification tests
- All identification algorithms now work with the new consistent role naming

daehyun99 commented:
It looks good to me, and the tests pass on my local environment.
I think we can merge it once the GitHub Action tests pass.

ankurankan committed:
Merge branch 'dev' into feature/add-exposures-outcomes-properties

ankurankan committed:
Adds exposures and outcomes arguments to DAG and PDAG classes

ankurankan commented:
@MAVRICK-1 Sorry for the delay in the review; I was away last couple of weeks. But I think the PR looks good to me; I have also added a commit adding the `exposures` and `outcomes` arguments to `DAG` and `PDAG` init methods. Could you please have a look if that looks alright to you?

@daehyun99 @Spinachboul FYI.

ankurankan committed:
Merge branch 'dev' into feature/add-exposures-outcomes-properties

ankurankan commented:
Merging this. @MAVRICK-1 Thanks for the PR.
